plugins/outputs/iotdb/iotdb.go

```diff
@@ -0,0 +1,277 @@
+	var deviceidList []string
+	var measurementsList [][]string
+	var valuesList [][]interface{}
+	var dataTypesList [][]client.TSDataType
+	var timestampList []int64
+	var tagsList [][]*telegraf.Tag
+
+	for _, metric := range metrics {
+		// write `metric` to the output sink here
+		var tags []*telegraf.Tag
+		tags = append(tags, metric.TagList()...)
+		// deal with basic parameter
+		var keys []string
+		var values []interface{}
+		var dataTypes []client.TSDataType
+		for _, field := range metric.FieldList() {
+			datatype, value := s.getDataTypeAndValue(field.Value)
+			if datatype == client.UNKNOW {
+				return nil, fmt.Errorf("field %q has unknown datatype, values: %v", field.Key, field.Value)
+			}
+			keys = append(keys, field.Key)
+			values = append(values, value)
+			dataTypes = append(dataTypes, datatype)
+		}
+		// Convert timestamp into specified unit
+		ts, err := s.convertTimestampOfMetric(metric)
+		if err != nil {
+			return nil, err
+		}
+		timestampList = append(timestampList, ts)
+		// append all metric data of this record to lists
+		deviceidList = append(deviceidList, metric.Name())
+		measurementsList = append(measurementsList, keys)
+		valuesList = append(valuesList, values)
+		dataTypesList = append(dataTypesList, dataTypes)
+		tagsList = append(tagsList, tags)
+	}
+	rwt := &recordsWithTags{
+		DeviceIDList:     deviceidList,
+		MeasurementsList: measurementsList,
+		ValuesList:       valuesList,
+		DataTypesList:    dataTypesList,
+		TimestampList:    timestampList,
+		TagsList:         tagsList,
+	}
+	return rwt, nil
+}
+
+// modify recordsWithTags according to 'TreatTagsAs' Configuration
+func (s *IoTDB) modifyRecordsWithTags(rwt *recordsWithTags) error {
+	if s.TreatTagsAs == "fields" {
+		// method 1: treat Tag(Key:Value) as measurement
+		for index, tags := range rwt.TagsList { // for each record
+			for _, tag := range tags { // for each tag of this record, append it's Key:Value to measurements
+				datatype, value := s.getDataTypeAndValue(tag.Value)
+				if datatype != client.UNKNOW {
+					rwt.MeasurementsList[index] = append(rwt.MeasurementsList[index], tag.Key)
+					rwt.ValuesList[index] = append(rwt.ValuesList[index], value)
+					rwt.DataTypesList[index] = append(rwt.DataTypesList[index], datatype)
+				}
+			}
+		}
+		return nil
+	} else if s.TreatTagsAs == "device_id" {
+		// method 2: treat Tag(Key:Value) as subtree of device id
+		for index, tags := range rwt.TagsList { // for each record
+			subfix := ""
+			for _, tag := range tags { // for each tag, append it's Value
+				subfix = subfix + "." + tag.Value
+			}
+			rwt.DeviceIDList[index] = rwt.DeviceIDList[index] + subfix
+		}
+		return nil
+	}
+	// something go wrong. This configuration should have been checked in func Init().
+	errorMsg := fmt.Sprintf("IoTDB Configuration Error: unknown TreatTagsAs: %s", s.TreatTagsAs)
+	s.Log.Errorf(errorMsg)
+	return errors.New(errorMsg)
```

Comment:
Please use a switch statement here and just return the error.
```suggestion
	switch s.TreatTagsAs {
	case "fields":
		// method 1: treat Tag(Key:Value) as measurement
		for index, tags := range rwt.TagsList { // for each record
			for _, tag := range tags { // for each tag of this record, append it's Key:Value to measurements
				datatype, value := s.getDataTypeAndValue(tag.Value)
				if datatype != client.UNKNOW {
					rwt.MeasurementsList[index] = append(rwt.MeasurementsList[index], tag.Key)
					rwt.ValuesList[index] = append(rwt.ValuesList[index], value)
					rwt.DataTypesList[index] = append(rwt.DataTypesList[index], datatype)
				}
			}
	case "device_id":
		// method 2: treat Tag(Key:Value) as subtree of device id
		for index, tags := range rwt.TagsList { // for each record
			var values []string
			for _, tag := range tags { // for each tag, append it's Value
				values = append(values, tag.Value)
			}
			rwt.DeviceIDList[index] = rwt.DeviceIDList[index] + strings.Join(values, ".")
		}
	default: 
		return fmt.Errorf("unknown  'convert_tags_to' option %q", s.TreatTagsAs)
	}
```

Reply:
all right.